Allow DC metadata to be entered in transfer, persist to ingest

mistydemeo opened the pull request:
These commits allow DC metadata to be entered in the transfer phase. This is implemented by adding a new DC form, which exists in the transfer phase, and saving the metadata to disk in between transfer and ingest.

To save and load DC metadata, the data is serialized to disk as JSON at the very end of a transfer using the new `saveDublinCore` microservice. This happens after the user selects to create a SIP, so it's unlikely that users would be able to try to manually edit metadata after it runs.

At the very beginning of ingest, immediately after the SIP is verified, the new `loadDublinCore` microservice looks for a `metadata/dc.json` file inside the SIP; if present, it populates the DC metadata for the SIP in the database using that file. This allows the DC metadata to be editable by hand during ingest, and makes sure it will be used in the final AIP METS.

~~To support this, this now persists the contents of the transfer's `metadata/` directory when creating a SIP. In the past, anything there would be discarded when creating a single SIP.~~

This PR's commit history is still p. messy, will get cleaned up on merge.

Commit:
Transfer detail view: enable DC metadata

This enables Dublin Core metadata to be entered in a transfer, as well
as in ingest. Later commits will ensure this is persisted into ingest
as expected.

refs #6217

Commit:
Add saveDublinCore microservice

This microservice serializes DC metadata to disk at the end of a
transfer, ensuring it's saved when the database is no longer present
(for instance, during SIP arrange).

refs #6217

Commit:
Add loadDublinCore microservice

This microservice loads DC metadata from disk and populates the DB at
the beginning of ingest. This allows DC metadata entered during the
transfer to be used the same way as DC metadata entered in ingest.
Since this happens at the beginning of ingest, it also ensures that
it continues to be editable via the UI from very early on in ingest.

refs #6217

Commit:
createSIPfromTransferObjects: copy JSON metadata

refs #6217

mistydemeo commented on src/MCPClient/lib/clientScripts/loadDublinCore.py:
```diff
@@ -0,0 +1,36 @@
+    dc = models.DublinCore(metadataappliestoidentifier=sip_uuid,
+                           metadataappliestotype=INGEST_METADATA_TYPE)
+    for key, value in data.iteritems():
+        setattr(dc, key, value)
```
Should probably do error checking in case users manually place `dc.json` files with invalid fields.

mistydemeo commented on src/MCPClient/lib/clientScripts/saveDublinCore.py:
```diff
@@ -0,0 +1,43 @@
+        return 0
+    for field in FIELDS:
+        attr = getattr(dc, field)
+        if attr:
```
Can `getattr()` fail to return a result here? Not sure how Django models would deal w/ that. This does at least prevent writing empty strings to the JSON.

Hwesta replied:
getattr raises an exception if no default value is provided - you probably want `getattr(dc, field, None)` or to handle the exception.  Using `models.DublinCore._meta.get_all_field_names()` instead of `FIELDS` might be useful as well, since it'll never return a field that doesn't exist (though you may not want all the fields - i.e. is_part_of (related to AIC work) isn't listed, and I assume that's intentional?

mistydemeo replied:
Yes, I left out `is_part_of` here because I only want the standard DC fields, not our extensions. `models.DublinCore._meta.get_all_field_names()` includes a few other undesirable fields, too (`id`, `metadataappliestoidentifier`, `metadataappliestotype`).

Commit:
Transfer DC metadata form: Remove unnecessary dropdown

refs #6480

Hwesta commented on src/MCPClient/lib/clientScripts/createSIPfromTransferObjects.py:
```diff
@@ -63,6 +63,15 @@
     for item in os.listdir(objectsDirectory):
         shutil.move(os.path.join(objectsDirectory, item), os.path.join(tmpSIPDir, "objects", item))
+    # Move the metadata/ directory, too
+    metadata_dir = os.path.join(objectsDirectory, '..', 'metadata')
```
`os.path.normpath` might be useful here.

Hwesta commented on src/MCPClient/lib/clientScripts/loadDublinCore.py:
```diff
@@ -0,0 +1,39 @@
+os.environ['DJANGO_SETTINGS_MODULE'] = 'settings.common'
+from main import models
+
+INGEST_METADATA_TYPE = '3e48343d-e2d2-4956-aaa3-b54d26eb9761'
```
Add a comment indicating this is the UUID of 'SIP' from MetadataAppliesToTypes table, please.

Hwesta commented on src/MCPClient/lib/clientScripts/loadDublinCore.py:
```diff
@@ -0,0 +1,39 @@
+def main(sip_uuid, dc_path):
+    # If there's no metadata, that's not an error, and just keep going
+    if not os.path.exists(dc_path):
+        print "DC metadata not found; exiting", "(at", dc_path, ")"
```
Why the extra `", "` between 'exiting' and '(at'?

Hwesta commented on src/dashboard/src/components/transfer/views.py:
```diff
@@ -141,3 +143,56 @@ def delete(request, uuid):
+                metadataappliestoidentifier=uuid
+            )
+
+    fields = ['title', 'creator', 'subject', 'description', 'publisher',
```
Since we've defined the fields list a few times (here, saveDublinCore.py, ingest.forms.AICDublinCoreMetadataForm/DublinCoreMetadataForm), maybe we should make it an attribute of DublinCore model?  Several of the fields lists vary on whether they include is_part_of or not, though.

mistydemeo replied:
I think a `DC_TERMS` that includes the standard DC terms set would make sense; `is_part_of` should only get added on a case-by-case basis.

Hwesta replied:
Sounds good, and that can be used in places where `_meta.get_all_field_names()` returns the extras we don't want.

Hwesta commented:
It seems a shame to make almost identical templates for metadata_{edit,list}.html, but there's already a precedent set with microservices.html.  Not sure of an easy way to fix it, though, other than passing in all the links that differ, and that gets awkward quickly.

mistydemeo commented:
Squashed and rebased.

I have no memory of why 223b8118838df063c648c1bc4305dc239b95476f is in this branch, but the method is no longer called anywhere so I have my suspicions it's totally unnecessary.

mistydemeo commented:
The only place it was called was this entry in urls, which is commented out: https://github.com/artefactual/archivematica/blob/qa/1.x/src/dashboard/src/components/filesystem_ajax/urls.py#L25 And has been since August: https://github.com/artefactual/archivematica/commit/4186da91d241d41c8e0972e97478a4b0d8072962#diff-222af4b335ab77a072b7bd5d18ed3960L23

mistydemeo commented:
Repushed without that commit.
